epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.kt

```diff
@@ -0,0 +1,208 @@
+        @IntRange(from = 0, to = 100) thresholdPercentage: Int
+    ) {
+        val previousPartiallyVisible = partiallyVisible
+        partiallyVisible = !detachEvent && isPartiallyVisible(thresholdPercentage)
+        if (partiallyVisible != previousPartiallyVisible) {
+            if (partiallyVisible) {
+                epoxyHolder.visibilityStateChanged(VisibilityState.PARTIAL_IMPRESSION_VISIBLE)
+            } else {
+                epoxyHolder.visibilityStateChanged(VisibilityState.PARTIAL_IMPRESSION_INVISIBLE)
+            }
+        }
+    }
+
+    fun handleFullImpressionVisible(epoxyHolder: EpoxyViewHolder, detachEvent: Boolean) {
+        val previousFullyVisible = fullyVisible
+        fullyVisible = !detachEvent && isFullyVisible()
+        if (fullyVisible != previousFullyVisible) {
+            if (fullyVisible) {
+                epoxyHolder.visibilityStateChanged(VisibilityState.FULL_IMPRESSION_VISIBLE)
+            }
+        }
+    }
+
+    fun handleChanged(epoxyHolder: EpoxyViewHolder, visibilityChangedEnabled: Boolean): Boolean {
+        var changed = false
+        if (visibleHeight != lastVisibleHeightNotified || visibleWidth != lastVisibleWidthNotified || viewVisibility != lastVisibilityNotified) {
+            if (visibilityChangedEnabled) {
+                if (viewVisibility == View.GONE) {
+                    epoxyHolder.visibilityChanged(0f, 0f, 0, 0)
+                } else {
+                    epoxyHolder.visibilityChanged(
+                        100f / height * visibleHeight,
+                        100f / width * visibleWidth,
+                        visibleHeight, visibleWidth
+                    )
+                }
+            }
+            lastVisibleHeightNotified = visibleHeight
+            lastVisibleWidthNotified = visibleWidth
+            lastVisibilityNotified = viewVisibility
+            changed = true
+        }
+        return changed
+    }
+
+    private fun isVisible(): Boolean {
+        return viewVisibility == View.VISIBLE && visibleHeight > 0 && visibleWidth > 0
+    }
+
+    // The model has entered the focused range either if it is larger than half of the viewport
+    // and it occupies at least half of the viewport or if it is smaller than half of the viewport
+    // and it is fully visible.
+    private val isInFocusVisible: Boolean
+        get() {
+            val halfViewportArea = viewportHeight * viewportWidth / 2
+            val totalArea = height * width
+            val visibleArea = visibleHeight * visibleWidth
+            // The model has entered the focused range either if it is larger than half of the viewport
+            // and it occupies at least half of the viewport or if it is smaller than half of the viewport
+            // and it is fully visible.
+            return viewVisibility == View.VISIBLE &&
+                if (totalArea >= halfViewportArea) visibleArea >= halfViewportArea else totalArea == visibleArea
+        }
```

Comment:
Should we just make this a function?
```suggestion
    private fun isInFocusVisible(): Boolean {
        val halfViewportArea = viewportHeight * viewportWidth / 2
        val totalArea = height * width
        val visibleArea = visibleHeight * visibleWidth
        // The model has entered the focused range either if it is larger than half of the viewport
        // and it occupies at least half of the viewport or if it is smaller than half of the viewport
        // and it is fully visible.
        return viewVisibility == View.VISIBLE &&
            if (totalArea >= halfViewportArea) visibleArea >= halfViewportArea else totalArea == visibleArea
    }
```

Reply:
> Should we just make this a function?

Interesting, Kotlin conversion converted a java method to `val`, good catch.